documentation/correctness.adoc

```diff
@@ -0,0 +1,102 @@
+= Model correctness
+
+We can identify a number of levels of correctness of a model.
+Each level builds on the previous one, so achieving a certain level of correctness implies having achieved all previous levels.
+The levels are divided into two layers.
+
+.Model correctness explained in a diagram
+image::../resources/correctness-diagrams/levels.svg[]
+
+
+== Correctness of a serialization chunk
+
+A JSON serialization of a model -- the _serialization chunk_ -- is considered to be deserializable (precisely) if it's correct with respect to all these levels.
+
+=== Low-level syntactic
+
+The serialization is parsable as JSON.
+Note that “parsable” doesn't necessarily mean “syntax-correct”.
+E.g., trailing commas are not syntax-correct, but don't necessarily prevent the JSON from being parsable, nor deserializable.
+
+Incorrectness is to be reported in terms of the first location where parsing the JSON text failed unrecoverably.
+This is typically dependent on the precise JSON parser used.
+
+=== Structural syntactic
+
+The JSON conforms to link:../schemas/serialization.schema.json[the JSON Schema for the serialization format]footnote:[This file should match link:https://github.com/LIonWeb-org/lioncore-typescript/blob/main/schemas/generic-serialization.schema.json[the one in the `lioncore-typescript` repository].].
+The `serializationFormatVersion` property of the root object of the serialization indicates which version to use (or should do so).
+Currently, there's only one version, which is identified as: `"1"` (i.e., a string containing the integer 1).
+
+Note that JSON Schema isn't powerful enough to express more high-level, nor constraints of a referential nature on the serialization chunk.
+
+Incorrectness is to be reported in terms of e.g. a https://github.com/json-path/JsonPath[JsonPath] to the location(s) of failure(s).
+This is typically dependent on the precise JSON Schema validator framework/library/utility used.
+
+=== Hierarchical
+
+The JSON doesn't violate the following constraints.
+
+1. IDs of serialized nodes are unique (within the serialization).
+Violation is to be reported in terms of precise locations in the JSON file of all the serialized nodes with equal ID.
+2. The `parent` property of a serialized node matches the `id` of the containing node (whenever that containing node is present in the serialization chunk).
+Concretely: if a node `N` is a child of a node `P` then `N.parent` should equal `P`.
```

Comment:
`N` can also be an annotation of `P`.

Reply:
Right, I really should implement the annotations in TS as well, so I actually know what's going on there as well.